src/backend/distributed/transaction/relation_access_tracking.c

```diff
@@ -524,10 +662,340 @@ GetRelationAccessMode(Oid relationId, ShardPlacementAccessType accessType)
 bool
 ShouldRecordRelationAccess()
 {
-	if (IsTransactionBlock() || InCoordinatedTransaction())
+	if (EnforceForeignKeyRestrictions &&
+		(IsTransactionBlock() || InCoordinatedTransaction()))
 	{
 		return true;
 	}
 
 	return false;
 }
+
+
+/*
+ * CheckConflictingRelationAccesses is mostly a wrapper around
+ * HoldsConflictingLockWithReferencingRelations(). We're only interested in accesses
+ * to reference tables that are referenced via a foreign constraint by a
+ * hash distributed tables.
+ */
+static void
+CheckConflictingRelationAccesses(Oid relationId, ShardPlacementAccessType accessType)
+{
+	DistTableCacheEntry *cacheEntry = NULL;
+	Oid conflictingReferencingRelationId = InvalidOid;
+	ShardPlacementAccessType conflictingAccessType = PLACEMENT_ACCESS_SELECT;
+
+	if (!EnforceForeignKeyRestrictions || !IsDistributedTable(relationId))
+	{
+		return;
+	}
+
+	cacheEntry = DistributedTableCacheEntry(relationId);
+
+	if (!(cacheEntry->partitionMethod == DISTRIBUTE_BY_NONE &&
+		  cacheEntry->referencingRelationsViaForeignKey != NIL))
+	{
+		return;
+	}
+
+	if (HoldsConflictingLockWithReferencingRelations(relationId, accessType,
```

Comment:
Note for the future: it would be nice if we only considered writes to reference tables that touch columns in the foreign key.